### PR TITLE
Mutation chains to stored procedures

### PR DESCRIPTION
**Pull request by dleard**

### CODE REVIEW

File: app/containers/Products/ProductRowItemContainer.js [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import React from 'react';
-import {graphql, createFragmentContainer, commitMutation} from 'react-relay';
+import {graphql, createFragmentContainer} from 'react-relay';
 import {Form, Button, ButtonGroup, Col, Row, Modal} from 'react-bootstrap';
+import {saveProductMutation} from '../../mutations/product/saveProductMutation';
+import {editBenchmarkMutation} from '../../mutations/benchmark/editBenchmarkMutation';
+import {createBenchmarkMutation} from '../../mutations/benchmark/createBenchmarkMutation';
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]: Use default import?

File: app/containers/Products/ProductRowItemContainer.js [hunk 2]
```diff
@@ -12,55 +15,8 @@ import {Form, Button, ButtonGroup, Col, Row, Modal} from 'react-bootstrap';
...
-    ) {
-      updateProductByRowId(input: $input) {
-        product {
-          rowId
-        }
-      }
-    }
-  `;
 
+export const ProductRowItemContainer = props => {
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]: Having a named export that is the same name as the file will trigger a linter error when you write 
```js
import ProductRowItemContainer from 'containers/Products/ProductRowItemContainer';
```
Rename to `ProductRowItemComponent` and/or rename the file to `ProductRowItem` ?  It's in the `containers` folder already, no need to append `Container` to the name

File: app/containers/Products/ProductRowItemContainer.js [hunk 3]
```diff
@@ -81,37 +37,26 @@ export const ProductRowItemContainer = props => {
...
-          await editBenchmark(currentBenchmark.rowId, benchmarkPatch);
-        }
-      },
-      onError: err => console.error(err)
-    });
+    const response = await saveProductMutation(
+      props.relay.environment,
+      variables
+    );
+    console.log(response);
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]:
```suggestion
```

File: app/containers/Products/ProductRowItemContainer.js [hunk 4]
```diff
@@ -125,91 +70,41 @@ export const ProductRowItemContainer = props => {
...
-      onCompleted: response => {
-        console.log(response);
-      },
-      onError: err => console.error(err)
-    });
+    const response = await editBenchmarkMutation(
+      props.relay.environment,
+      variables
+    );
+    console.log(response);
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]:
```suggestion
```

File: app/containers/Products/ProductRowItemContainer.js [hunk 5]
```diff
@@ -125,91 +70,41 @@ export const ProductRowItemContainer = props => {
...
+    const variables = {
       input: {
-        product: {
-          name: event.nativeEvent.target[3].value,
-          description: event.nativeEvent.target[4].value,
-          state: 'active',
-          parent: [props.product.rowId]
-        }
+        newName: event.nativeEvent.target[3].value,
+        newDescription: event.nativeEvent.target[4].value,
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]: What are `3` and `4`. Named constants are better than magic numbers :)

File: app/containers/Products/ProductRowItemContainer.js [hunk 6]
```diff
@@ -125,91 +70,41 @@ export const ProductRowItemContainer = props => {
...
-          await editBenchmark(currentBenchmark.rowId, benchmarkPatch);
-        }
-      },
-      onError: err => console.error(err)
-    });
+    const response = await saveProductMutation(
+      props.relay.environment,
+      variables
+    );
+    console.log(response);
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]:
```suggestion
```

File: app/containers/Products/ProductRowItemContainer.js [hunk 7]
```diff
@@ -256,34 +147,26 @@ export const ProductRowItemContainer = props => {
...
-        benchmark: {
-          productId: props.product.rowId,
-          benchmark: parseFloat(event.nativeEvent.target[3].value),
-          eligibilityThreshold: parseFloat(event.nativeEvent.target[4].value),
-          startDate,
-          updatedAt: currentDate,
-          updatedBy: 'Admin'
-        }
+        productIdInput: props.product.rowId,
+        benchmarkInput: parseFloat(event.nativeEvent.target[3].value),
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]: 3!

File: app/containers/Products/ProductRowItemContainer.js [hunk 8]
```diff
@@ -256,34 +147,26 @@ export const ProductRowItemContainer = props => {
...
+        eligibilityThresholdInput: parseFloat(
+          event.nativeEvent.target[4].value
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]: 4!

File: app/containers/Products/ProductRowItemContainer.js [hunk 9]
```diff
@@ -256,34 +147,26 @@ export const ProductRowItemContainer = props => {
...
-          await editBenchmark(currentBenchmark.rowId, benchmarkPatch);
-        }
-      },
-      onError: err => console.error(err)
-    });
+    const response = await createBenchmarkMutation(
+      props.relay.environment,
+      newVariables
+    );
+    console.log(response);
```
Comment (matthieu-foucault) [inline app/containers/Products/ProductRowItemContainer.js]:
```suggestion
```

File: app/mutations/benchmark/createBenchmarkMutation.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        benchmark
+        eligibilityThreshold
+        startDate
+        endDate
+      }
+    }
+  }
+`;
+
+export const createBenchmarkMutation = async (environment, variables) => {
```
Comment (matthieu-foucault) [inline app/mutations/benchmark/createBenchmarkMutation.js]: export default?

File: app/mutations/benchmark/editBenchmarkMutation.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+        id
+      }
+    }
+  }
+`;
+
+// TODO: abstract clientMutationId into a base class
+// TODO: May want to surface the onCompleted errors to the user (ie not reject, resolve & report)
+// TODO: Add optimistic updates https://relay.dev/docs/en/mutations https://relay.dev/docs/en/relay-store
+export const editBenchmarkMutation = async (environment, variables) => {
```
Comment (matthieu-foucault) [inline app/mutations/benchmark/editBenchmarkMutation.js]: export default?

File: app/mutations/product/saveProductMutation.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+          }
+        }
+      }
+    }
+  }
+`;
+
+// TODO: abstract clientMutationId into a base class
+// TODO: May want to surface the onCompleted errors to the user (ie not reject, resolve & report)
+export const saveProductMutation = async (environment, variables) => {
```
Comment (matthieu-foucault) [inline app/mutations/product/saveProductMutation.js]: default export?

File: app/containers/Products/ProductRowItemContainer.js (hunk 2 above)
Comment (dleard) [reply app/containers/Products/ProductRowItemContainer.js]: A review & shared understanding on naming conventions would be helpful for everyone on the team. I think we are all doing different things.

Comment (dleard) [reply app/containers/Products/ProductRowItemContainer.js]: Changing the name of the function broke all sorts of stuff, possibly to do with the export / default export we are doing in this file. I've changed it back for now & maybe we can revisit how to change these after we have defined a style guide.

File: app/containers/Products/ProductRowItemContainer.js (hunk 3 above)
Comment (dleard) [reply app/containers/Products/ProductRowItemContainer.js]: placeholder for logging / displaying errors to the user

File: app/mutations/benchmark/createBenchmarkMutation.js (hunk 10 above)
Comment (dleard) [reply app/mutations/benchmark/createBenchmarkMutation.js]: default added

File: app/mutations/benchmark/editBenchmarkMutation.js (hunk 11 above)
Comment (dleard) [reply app/mutations/benchmark/editBenchmarkMutation.js]: default added

File: app/mutations/product/saveProductMutation.js (hunk 12 above)
Comment (dleard) [reply app/mutations/product/saveProductMutation.js]: default added

**Pull request merged**